File: cspro-dev/cspro/CSEntryDroid/app/src/main/java/gov/census/cspro/bridge/CNPifFile.java
package gov.census.cspro.bridge;

import android.os.Bundle;

import java.util.ArrayList;

import gov.census.cspro.engine.Util;

public class CNPifFile
{
	public enum ShowInApplicationListing { Always, Hidden, Never };
	
	private boolean valid;
	private String description;
	private boolean entryAppType;
	private ShowInApplicationListing showInApplicationListing;
	private String inputFilename;
	private String appFilename;
	private String[] externalFilenames;
	private String[] userFilenames;
	private String writeFilename;
	private String onExitFilename;
	private String pffDirectory;
	
	protected native long LoadPif(String filename);
	protected native String GetDescription(long reference);
	protected native boolean IsAppTypeEntry(long reference);
	protected native int GetShowInApplicationListing(long reference);
	protected native String GetInputFilename(long reference);
	protected native String GetAppFilename(long reference);
	protected native void ClosePif(long reference);
	protected native String[] GetExternalFilenames(long reference);
	protected native String[] GetUserFilenames(long reference);
	protected native String GetWriteFilename(long reference);
	protected native String GetOnExitFilename(long reference);
	
	public CNPifFile(String filename)
	{
		long reference = LoadPif(filename);
		
		valid = ( reference != 0 );
		
		if( valid )
		{
			pffDirectory = Util.removeFilename(filename);
			description = GetDescription(reference);
            entryAppType = IsAppTypeEntry(reference);
			showInApplicationListing = ShowInApplicationListing.values()[GetShowInApplicationListing(reference)];
			inputFilename = GetInputFilename(reference);
			appFilename = GetAppFilename(reference);
			externalFilenames = GetExternalFilenames(reference);
			userFilenames = GetUserFilenames(reference);
			writeFilename = GetWriteFilename(reference);
			onExitFilename = GetOnExitFilename(reference);
			ClosePif(reference);
		}
	}
	
	public boolean IsValid()
	{
		return valid;
	}
	
	public String GetDescription()
	{
		return description;		
	}

	public Boolean IsAppTypeEntry()
	{
		return entryAppType;
	}

	public boolean ShouldShowInApplicationListing(boolean showHiddenApplications)
	{
		// if the description is empty then we won't display it in the Applications Listing screen
		return ( description.trim().length() > 0 ) && ( ( showInApplicationListing == ShowInApplicationListing.Always ) ||
				( showHiddenApplications && ( showInApplicationListing == ShowInApplicationListing.Hidden ) ) );		
	}
	
	public String GetInputFilename()
	{
		return inputFilename;		
	}
	
	public String GetAppFilename()
	{
		return appFilename;
	}
	
	public String[] GetExternalFilenames()
	{
		return externalFilenames;
	}

	public String[] GetUserFilenames()
	{
		return userFilenames;
	}

	public String GetWriteFilename()
	{
		return writeFilename;
	}
	
	public String GetOnExitFilename()
	{
		return onExitFilename;
	}

    public native static String CreatePffFromIntentExtras(String pff_file, Bundle bundle_extras);
}

File: src/commonMain/kotlin/gov/census/cspro/bridge/CNPifFile.kt
package gov.census.cspro.bridge

import gov.census.cspro.util.Util

/**
 * CSPro .pff (Program Information File) parser
 * Ported from Android CNPifFile.java
 * 
 * Uses expect/actual pattern for native PFF parsing
 */
class CNPifFile private constructor() {
    
    enum class ShowInApplicationListing {
        Always,
        Hidden,
        Never
    }
    
    var isValid: Boolean = false
        private set
    
    var description: String = ""
        private set
    
    var entryAppType: Boolean = false
        private set
    
    var showInApplicationListing: ShowInApplicationListing = ShowInApplicationListing.Always
        private set
    
    var inputFilename: String = ""
        private set
    
    var appFilename: String = ""
        private set
    
    var externalFilenames: Array<String> = emptyArray()
        private set
    
    var userFilenames: Array<String> = emptyArray()
        private set
    
    var writeFilename: String = ""
        private set
    
    var onExitFilename: String = ""
        private set
    
    var pffDirectory: String = ""
        private set
    
    fun shouldShowInApplicationListing(showHiddenApplications: Boolean): Boolean {
        // if the description is empty then we won't display it in the Applications Listing screen
        return description.trim().isNotEmpty() && (
            showInApplicationListing == ShowInApplicationListing.Always ||
            (showHiddenApplications && showInApplicationListing == ShowInApplicationListing.Hidden)
        )
    }
    
    private fun setFromData(filename: String, data: CNPifFileData) {
        isValid = data.isValid
        if (isValid) {
            pffDirectory = Util.removeFilename(filename)
            description = data.description
            entryAppType = data.entryAppType
            showInApplicationListing = ShowInApplicationListing.values()[data.showInApplicationListing]
            inputFilename = data.inputFilename
            appFilename = data.appFilename
            externalFilenames = data.externalFilenames
            userFilenames = data.userFilenames
            writeFilename = data.writeFilename
            onExitFilename = data.onExitFilename
        }
    }
    
    companion object {
        /**
         * Synchronous loader - for use when PFF content is already cached
         */
        fun load(filename: String): CNPifFile {
            val pif = CNPifFile()
            val data = CNPifFileLoader.load(filename)
            pif.setFromData(filename, data)
            return pif
        }
        
        /**
         * Asynchronous loader - for loading from OPFS
         */
        suspend fun loadAsync(filename: String): CNPifFile {
            val pif = CNPifFile()
            val data = CNPifFileLoader.loadAsync(filename)
            pif.setFromData(filename, data)
            return pif
        }
    }
}

/**
 * Data class for PFF file contents
 */
data class CNPifFileData(
    val isValid: Boolean = false,
    val description: String = "",
    val entryAppType: Boolean = false,
    val showInApplicationListing: Int = 0,
    val inputFilename: String = "",
    val appFilename: String = "",
    val externalFilenames: Array<String> = emptyArray(),
    val userFilenames: Array<String> = emptyArray(),
    val writeFilename: String = "",
    val onExitFilename: String = ""
)

/**
 * Platform-specific PFF loader
 */
expect object CNPifFileLoader {
    fun load(filename: String): CNPifFileData
    suspend fun loadAsync(filename: String): CNPifFileData
}
